File: Python DSA/StackQueue/evaluate_reverse_polish_notation.py
"""
EVALUATE REVERSE POLISH NOTATION - Medium
Evaluate RPN expression.

Logical Thinking:
1. Use stack
2. Push numbers
3. Pop two operands for operators

Logical Thinking:
def eval_rpn(tokens):
    stack = []
    for token in tokens:
        if token in '+-*/':
            b, a = stack.pop(), stack.pop()
            if token == '+':
                stack.append(a + b)
            elif token == '-':
                stack.append(a - b)
            elif token == '*':
                stack.append(a * b)
            else:
                stack.append(int(a / b))
        else:
            stack.append(int(token))
    return stack[0]
"""

# TODO: Implement solution


# Test cases
if __name__ == "__main__":
    print(f"Testing Evaluate Reverse Polish Notation...")
    print("Add your test cases here")

File: DSA/Stack/Java/EvaluateReversePolishNotation.java
// LeetCode 150: Evaluate Reverse Polish Notation
// Evaluate the value of an arithmetic expression in Reverse Polish Notation.

import java.util.Stack;

public class EvaluateReversePolishNotation {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();
        
        for (String token : tokens) {
            if (isOperator(token)) {
                int b = stack.pop();
                int a = stack.pop();
                int result = performOperation(a, b, token);
                stack.push(result);
            } else {
                stack.push(Integer.parseInt(token));
            }
        }
        
        return stack.pop();
    }
    
    private boolean isOperator(String token) {
        return token.equals("+") || token.equals("-") || 
               token.equals("*") || token.equals("/");
    }
    
    private int performOperation(int a, int b, String operator) {
        switch (operator) {
            case "+": return a + b;
            case "-": return a - b;
            case "*": return a * b;
            case "/": return a / b;
            default: throw new IllegalArgumentException("Invalid operator: " + operator);
        }
    }
    
    public static void main(String[] args) {
        EvaluateReversePolishNotation solution = new EvaluateReversePolishNotation();
        
        String[] tokens1 = {"2", "1", "+", "3", "*"};
        String[] tokens2 = {"4", "13", "5", "/", "+"};
        
        System.out.println("Result 1: " + solution.evalRPN(tokens1)); // 9
        System.out.println("Result 2: " + solution.evalRPN(tokens2)); // 6
    }
}